Use bookworm; target specific Python & PostgreSQL

AndrewFerr opened the pull request:
Update dependencies to comply with Synapse's deprecation policy: https://element-hq.github.io/synapse/latest/deprecation_policy.html

- Upgrade to Debian Bookworm (the current oldstable) to pull in the oldest supported version of SQLite.
- Instead of installing PostgreSQL from Debian repositories, use the PostgreSQL Apt Repository to install its lowest-supported version.
- Instead of installing Python from Debian repositories, use uv to install its lowest-supported version.

AndrewFerr committed:
Use bookworm; target specific Python & PostgreSQL

Update dependencies to comply with Synapse's deprecation policy:
https://element-hq.github.io/synapse/latest/deprecation_policy.html

- Upgrade to Debian Bookworm (the current oldstable) to pull in the
  oldest supported version of SQLite.
- Instead of installing PostgreSQL from Debian repositories, use the
  PostgreSQL Apt Repository to install its lowest-supported version.
- Instead of installing Python from Debian repositories, use uv to
  install its lowest-supported version.

AndrewFerr commented:
The CI tests depend on new `sytest` and `sytest-synapse` images to be built, as the tests run in containers of the published `matrixdotorg/sytest-*` images.  Is there a good way to use freshly-built images from the current PR instead?

-> https://github.com/matrix-org/sytest/pull/1414#discussion_r2452838083

AndrewFerr committed:
Move build arg to correct location

AndrewFerr committed:
Install postgresql just after other installations

This is what was done before, and it affects the installation of DBD::Pg

AndrewFerr committed:
Merge branch 'develop' into af/bookworm

AndrewFerr committed:
Clean apt lists after every apt install

to minimize build layer size

MadLittleMods commented on docker/base.Dockerfile:
```diff
@@ -20,14 +20,19 @@ RUN apt-get -qq update && apt-get -qq install -y \
     && rm -rf /var/lib/apt/lists/*
 
+ARG POSTGRESQL_VERSION=13
+RUN /usr/share/postgresql-common/pgdg/apt.postgresql.org.sh -y \
+    && apt -qq install -y postgresql-${POSTGRESQL_VERSION} \
+    && rm -rf /var/lib/apt/lists/*
```
> Instead of installing PostgreSQL from Debian repositories, use the PostgreSQL Apt Repository to install its lowest-supported version.

Do we do this anywhere else? I'd be more keen if this magic was already accepted elsewhere

AndrewFerr replied:
To my knowledge, we don't do it elsewhere, but this is the recommended strategy from upstream PostgreSQL docs for installing a specific version in Debian: https://www.postgresql.org/download/linux/debian/

AndrewFerr replied:
Ah, just found that it's also recommended in the Debian Wiki, as well: https://wiki.debian.org/PostgreSql%20%20#PGDG_Repository

MadLittleMods replied:
Let's cross-link those references in a comment :fast_forward: 

We should also explain why we care about installing that minimum version (the Synapse deprecation policy)

AndrewFerr replied:
4e15850c for repo usage

0c73ed7b for min version

MadLittleMods replied:
We currently only link one of the references. Are both references useful? I haven't inspected them too closely but if I was tackling this task in the future, I would want to at-least look at both of these so having them to jump from quickly seems good to me.

AndrewFerr replied:
The Debian Wiki page references a PostgreSQL wiki page, which has similar information to the docs page on www.postgresql.org, so I figured it was more efficient to link just the Debian Wiki page.

MadLittleMods commented on docker/synapse.Dockerfile:
```diff
@@ -21,10 +18,13 @@ RUN mkdir /rust /cargo
+
+RUN --mount=type=bind,from=ghcr.io/astral-sh/uv:0.9.4,source=/uv,target=/bin/uv \
+        uv python install "$PYTHON_VERSION" && \
+        uv tool install poetry@1.3.2
```
> Instead of installing Python from Debian repositories, use uv to install its lowest-supported version.

Do we do this anywhere else? I'd be more keen if this magic was already accepted elsewhere

AndrewFerr replied:
Not sure if we use this exact strategy elsewhere, but I got the idea of doing this from Synapse's (relatively) recent usage of `uv` for its Docker image builds.

MadLittleMods replied:
How do we handle this in Synapse for other things?

In CI, it looks like we set `python-version` and `matrix-org/setup-python-poetry` takes care of things.

In our Docker images, it looks like we use an image directly that bundles it in: [`FROM ghcr.io/astral-sh/uv:python${PYTHON_VERSION}-${DEBIAN_VERSION}`](https://github.com/element-hq/synapse/blob/cba3a814c63ad877c482df3bc75570e4a7d61ddb/docker/Dockerfile#L68)

Not sure how much effort it would be to just use the direct image that looks like it already using Debian and has the specified Python version bundled in.

It looks relatively sane to do it the way you have it :fast_forward:

AndrewFerr replied:
The `uv:python3.X` images are derived from a base distribution, whereas the `uv:X.Y.Z` ones are distroless images that include just the `uv` binaries ([source](https://docs.astral.sh/uv/guides/integration/docker/#getting-started)).

Synapse's CI uses `uv` during its build steps, so it benefits from using a derived image with `uv` available as a build layer image base.

For sytest's CI, I thought it made more sense to use `uv` solely to "install" a specific Python version.

MadLittleMods replied:
I'm not following completely as we currently use Debian and `uv` to install a certain Python version. And it looks like we can use Debian-based images from `uv` with the Python version directly.

In any case, prior art :fast_forward:

MadLittleMods commented on docker/synapse.Dockerfile:
```diff
@@ -21,10 +18,13 @@ RUN mkdir /rust /cargo
+ARG PYTHON_VERSION=python3
+
+RUN --mount=type=bind,from=ghcr.io/astral-sh/uv:0.9.4,source=/uv,target=/bin/uv \
+        uv python install "$PYTHON_VERSION" && \
```
> Instead of installing Python from Debian repositories, use uv to install its lowest-supported version.

How does `uv python install python3` end up installing the minimum supported Python version which should be Python 3.9?

MadLittleMods replied:
Ahh, I think `ARG PYTHON_VERSION=python3` is the default and we override in the CI matrix. See `PYTHON_VERSION=python3.9` above for example.

AndrewFerr replied:
Exactly.  Since that caused some confusion, I'll refactor where/how this variable is set to make it more clear.

AndrewFerr replied:
Done with a94a67bb

MadLittleMods commented on .github/workflows/docker.yml:
```diff
@@ -23,10 +23,12 @@ jobs:
     strategy:
```
> The CI tests depend on new `sytest` and `sytest-synapse` images to be built, as the tests run in containers of the published `matrixdotorg/sytest-*` images.  Is there a good way to use freshly-built images from the current PR instead?

I'm not sure as I never deal with Sytest.

It looks like @H-Shay did this last -> https://github.com/matrix-org/sytest/pull/1358 - Any insight here?

H-Shay replied:
Hmm it's been quite awhile and I didn't do this regularly, but I think there was a way to manually trigger the build - https://github.com/matrix-org/sytest/pull/1358#issuecomment-1626054011 
I don't have the same access I did then so I can't look but I think there was a run button on one of the actions that could trigger this.

MadLittleMods replied:
On https://github.com/matrix-org/sytest/actions/workflows/pipeline.yml, I see a "Build and deploy docker images" workflow to run manually but not for a branch that is in a fork (as far I as I can tell).

I ended up doing this:

 1. `git fetch origin refs/pull/1414/head:pr-1414`
 1. `git checkout pr-1414`
 1. `git push --set-upstream origin pr-1414`
 1. On https://github.com/matrix-org/sytest/actions/workflows/docker.yml, run the "Build and deploy docker images" workflow for the `pr-1414` branch
 1. -> https://github.com/matrix-org/sytest/actions/runs/18729194682
  Looks like it mostly passed:
    - `Build sytest:bookworm` :white_check_mark: 
    - `Build sytest:testing` :white_check_mark: 
    - `Build sytest-synapse:bookworm` (canceled, why? -> `The strategy configuration was canceled because "build-dependent-images.testing_synapse_matrixdot" failed`)
    - `Build sytest-synapse:testing` :x:
    ```
    Error: buildx failed with: ERROR: failed to build: failed to solve: process "/bin/bash -o pipefail -c wget -q https://github.com/element-hq/synapse/archive/develop.tar.gz         -O /synapse.tar.gz &&         mkdir /synapse &&         tar -C /synapse --strip-components=1 -xf synapse.tar.gz &&         ln -s -T /venv /synapse/.venv &&         cd /synapse &&         poetry install -q --no-root --extras all &&         rm -rf `poetry config cache-dir` &&         rm -rf /synapse &&         rm /synapse.tar.gz" did not complete successfully: exit code: 1
    ```
     
 Perhaps we just need to re-run the CI here now - or fix up the builds and run them again.

AndrewFerr replied:
The failure is from `poetry install`, but the details are suppressed by `-q`. Rebuilding locally without it shows these details:
```
error: the configured Python interpreter version (3.14) is newer than PyO3's maximum supported version (3.13)
= help: please check if an updated version of PyO3 is available. Current version: 0.24.1
= help: set PYO3_USE_ABI3_FORWARD_COMPATIBILITY=1 to suppress this check and build anyway using the stable ABI
```

Addressed by 1ab7a300

MadLittleMods replied:
New rebuild underway -> https://github.com/matrix-org/sytest/actions/runs/18752219369

AndrewFerr replied:
It passed :partying_face:

MadLittleMods replied:
CI is all green here as well :white_check_mark: :tada: 

@AndrewFerr Should I merge now?

AndrewFerr replied:
Yes, please!

MadLittleMods commented on .github/workflows/docker.yml:
```diff
@@ -23,10 +23,12 @@ jobs:
-            tag: bullseye
+          - base_image: debian:bookworm
+            tag: bookworm
+            postgresql_version: 13
```
How can I cross-verify these version numbers?

AndrewFerr replied:
https://endoflife.date/postgresql, which I found from [Synapse's deprecation policy docs](https://element-hq.github.io/synapse/latest/deprecation_policy.html).

The Debian tag used here is the codename of whatever `oldstable` is at the time of writing.  Maybe it'd be more convenient to use `oldstable`, but pinning on a specific version is more reliable.

For the Debian-provided version of PostgreSQL, see https://packages.debian.org/oldstable/postgresql, though that is less relevant when installing from the upstream PostgreSQL repo.

AndrewFerr committed:
Use min-supported python & postgresQL by default

Also refactor the CI matrix for the base image to use the same format as
that of the dependent images, and make PYTHON_VERSION purely numeric

MadLittleMods commented on docker/synapse.Dockerfile:
```diff
@@ -21,10 +18,13 @@ RUN mkdir /rust /cargo
-# installation.
-# Some context in https://github.com/pypa/pip/issues/11381#issuecomment-1399263627
-RUN ${PYTHON_VERSION} -m pip install -q --no-cache-dir poetry==1.3.2 ${SYSTEM_PIP_INSTALL_SUFFIX}
+ARG PYTHON_VERSION=3.9
```
I'm thinking we should maybe add a comment about this default being the minimum Python version supported according to the [Synapse deprecation policy](https://element-hq.github.io/synapse/latest/deprecation_policy.html).

Especially since this is documented in another project.

AndrewFerr replied:
0c73ed7b

MadLittleMods commented on .github/workflows/docker.yml:
```diff
@@ -63,16 +65,16 @@ jobs:
-            tags: "matrixdotorg/sytest-synapse:bullseye"
-            build_args: "SYTEST_IMAGE_TAG=bullseye"
+            tags: "matrixdotorg/sytest-synapse:bookworm"
+            build_args: "SYTEST_IMAGE_TAG=bookworm"
```
```suggestion
            build_args: |
              SYTEST_IMAGE_TAG=bookworm
              # This is the default from the image
              # PYTHON_VERSION=3.9
```

(more obvious)

AndrewFerr replied:
If showing the default `PYTHON_VERSION` here is more obvious, would it be better to not give it a default at all, and rely on it being set in every CI matrix value?  (This applies to the `POSTGRES_VERSION` build arg too)

MadLittleMods replied:
Depends what perspective you're coming from. For CI, that makes sense.

What's the usual developer flow for running Sytest locally? If it works without more configuration for local developers, that sounds fine. I recall an option to run Sytest in Docker so it would be good to use the latest Python version in that case.

AndrewFerr replied:
Presumably, developers wanting to run Sytest in Docker would just pull our prebuilt Sytest images instead of building them manually.  And anyone developing Sytest itself would probably run it natively.  So it should be rare for developers to need to build Sytest images & have to set missing build args themselves.

Though out of principle, I generally prefer setting good `ARG` defaults, especially since these Dockerfiles already give defaults to some other important `ARG`s, like `BASE_IMAGE` and `SYTEST_IMAGE_TAG`.

...but with that said, CI always sets `BASE_IMAGE` and `SYTEST_IMAGE_TAG` even when their defaults would suffice.  So for consistency's sake, I'll make CI always set the `*_VERSION` args too.

AndrewFerr replied:
16383797

MadLittleMods commented on docker/synapse.Dockerfile:
```diff
@@ -21,10 +18,13 @@ RUN mkdir /rust /cargo
-# installation.
-# Some context in https://github.com/pypa/pip/issues/11381#issuecomment-1399263627
-RUN ${PYTHON_VERSION} -m pip install -q --no-cache-dir poetry==1.3.2 ${SYSTEM_PIP_INSTALL_SUFFIX}
+ARG PYTHON_VERSION=3.9
```
I'm not sure if the minimum Python version is the best default to have 🤷 

Since this is just for testing, it's probably fine.

MadLittleMods replied:
As I suspected, we use the [latest supported Python version ](https://github.com/element-hq/synapse/blob/cba3a814c63ad877c482df3bc75570e4a7d61ddb/docker/Dockerfile#L24) as the default in Synapse Docker image. But that's also used in production for people. And then we use that image to build the Complement image.

nico-famedly replied:
It might make sense to go to 3.10 directly though, considering that 3.9 is EOL in a week. Unless someone will bump it then. Otherwise you will get conflicts when the minimum version is bumped in Synapse.

AndrewFerr replied:
9cd8e8b6

AndrewFerr committed:
Drop max Python version to 3.13

as 3.14 is too new for PyO3, pulled in by poetry install when it builds
wheels for pydantic_core

AndrewFerr committed:
Don't run poetry install quietly

which will prevent suppressing build errors

AndrewFerr committed:
Document default Python & PostgreSQL versions

AndrewFerr committed:
Document upstream PostgreSQL repo usage

AndrewFerr committed:
Bump default Python version to 3.10

as that will be the minimum-supported version as of 31-01-2025

MadLittleMods commented on .github/workflows/pipeline.yml:
```diff
@@ -21,23 +21,26 @@ jobs:
         include:
-          - label: Debian Bullseye (Py 3.9, SQLite) Monolith
-            sytest-tag: bullseye
+          - label: Debian Bookworm (Py 3.9, SQLite 3.40) Monolith
```
The minimum version was bumped to `3.10` now (https://github.com/matrix-org/sytest/pull/1414#discussion_r2454981332)

We should update these labels

AndrewFerr replied:
Good catch. ebd42a60

AndrewFerr committed:
Update Python version in pipeline job labels

AndrewFerr committed:
Always set python & postgresql version args in CI

for consistency with how CI always sets the base image tag, despite all
of these build args having equivalent defaults to what is set
